File: Problems/0101. (E) Symmetric Tree/Python3/solution.py
class Solution:
  # recursive solution
  def isSymmetric(self, root: TreeNode) -> bool:
    return self.recurse(root.left, root.right)
  
  def recurse(self, left, right):
    # base cases
    if not left and not right:
      return True
    if not left or not right:
      return False
    
    # check current pair and compare recursively
    return left.val == right.val and
           self.recurse(left.left, right.right) and
           self.recurse(left.right, right.left) 

File: Problems/0101. (E) Symmetric Tree/Java/solution.java
class Solution {
  // recursive solution
  public boolean isSymmetric(TreeNode root) {
    return recurse(root.left, root.right);
  }

  private boolean recurse(TreeNode left, TreeNode right) {
    // base cases
    if (left == null && right == null)
      return true;
    if (left == null || right == null)
      return false;

    // check current pair and compare recursively
    return left.val == right.val &&
           recurse(left.left, right.right) &&
           recurse(left.right, right.left);
  }
}
